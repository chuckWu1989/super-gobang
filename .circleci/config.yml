defaults: &defaults
  docker:
    - image: circleci/node:7.10
  working_directory: ~/superGobang
version: 2
jobs:
  build:
    <<: *defaults
    steps:
      - checkout
      - restore_cache:
          keys:
          - v1-dependencies-{{ checksum "package.json" }}
          - v1-dependencies-
      - run: yarn install
      - save_cache:
          paths:
            - node_modules
          key: v1-dependencies-{{ checksum "package.json" }}
  test:
    <<: *defaults
    - attach_workspace:
      at: ~/superGobang
    steps:
      - checkout
      - run: yarn test

all_tags: &all_tags
  filters:
    branches:
      only: master
workflows:
  version: 2
  build_and_test:
    jobs:
      - build:
        <<: *all_tags
      - test:
        <<: *all_tags
        requires:
          - build
