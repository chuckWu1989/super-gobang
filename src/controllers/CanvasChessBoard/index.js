const drawLine = (ctx, s, e) => {
  ctx.beginPath();
  ctx.moveTo(s[0], s[1]);
  ctx.lineTo(e[0], e[1]);
  ctx.closePath();
  ctx.stroke();
};
export const drawBoard = (canvas, props) => {
  const {
    grid,
    coord,
    width,
    ratio,
  } = props;
  const ctx = canvas.getContext('2d');
  const height = width * ratio;
  ctx.clearRect(0, 0, width, height);
  for (let i = 0; i <= (grid + 1); i += 1) {
    drawLine(ctx, coord[0][i], coord[grid + 1][i]);
    drawLine(ctx, coord[i][0], coord[i][grid + 1]);
  }
};

export default drawBoard;
