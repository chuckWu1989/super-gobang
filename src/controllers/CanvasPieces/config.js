export const radiusRatio = 0.8;
export const s = 0 * Math.PI;
export const e = 2 * Math.PI;
