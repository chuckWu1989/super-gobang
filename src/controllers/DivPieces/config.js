export const radiusRatio = 0.8;

export default radiusRatio;
