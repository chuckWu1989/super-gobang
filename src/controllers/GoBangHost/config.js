export const chainNum = 5;
export const black = 'black';
export const white = 'white';
export const deadlock = 'deadlock';

export default chainNum;
