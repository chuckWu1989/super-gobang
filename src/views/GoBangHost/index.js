import React, { Component } from 'react';
import PropTypes from 'prop-types';
import ResponsiveContainer from '../../router/ResponsiveContainer';
import DGoBang from '../DivGoBang';
import CGoBang from '../CanvasGoBang';
import './style.less';

class GoBangHost extends Component {
  static defaultProps = {
    grid: 15,
    bColor: '#000000',
    wColor: '#e7e7eb',
    singleRace: false,
    fallback: false,
    restart: false,
    gameOver: () => {},
    updated: () => {},
    addClickEventListener: () => {},
    removeClickEventListener: () => {},
    receivePropsHandler: () => {},
    shouldUpdate: () => {},
    isCanvasSupported: () => true,
  }
  static propTypes = {
    grid: PropTypes.number,
    bColor: PropTypes.string,
    wColor: PropTypes.string,
    gameOver: PropTypes.func,
    updated: PropTypes.func,
    addClickEventListener: PropTypes.func,
    removeClickEventListener: PropTypes.func,
    receivePropsHandler: PropTypes.func,
    shouldUpdate: PropTypes.func,
    isCanvasSupported: PropTypes.func,
    singleRace: PropTypes.bool,
    fallback: PropTypes.bool,
    restart: PropTypes.bool,
  }
  constructor(props) {
    super(props);
    this.state = { chessMoves: [], WrappedComponent: undefined };
  }
  componentWillMount() {
    this.props.addClickEventListener.call(this, this.props);
    const supportCanvas = this.props.isCanvasSupported();
    const WrappedComponent = supportCanvas ? CGoBang : DGoBang;
    this.setState({ WrappedComponent });
  }
  componentWillReceiveProps(nextProps) {
    this.props.receivePropsHandler.call(this, nextProps);
  }
  shouldComponentUpdate(nextProps, nextState) {
    return this.props.shouldUpdate(nextState, this.state);
  }
  componentDidUpdate() {
    this.props.updated();
  }
  componentWillUnmount() {
    this.props.removeClickEventListener.call(this);
  }
  render() {
    const { chessMoves, WrappedComponent } = this.state;
    const { grid } = this.props;
    return (
      <div className="gobang-host-style">
        <ResponsiveContainer>
          <WrappedComponent grid={grid} chessMoves={chessMoves} />
        </ResponsiveContainer>
      </div>
    );
  }
}

export default GoBangHost;
