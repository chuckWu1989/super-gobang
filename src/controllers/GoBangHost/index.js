import { chainNum, black, white, deadlock } from './config';

/**
 * Check if the point of piece is in boundary of grid
 * @param {number} x The x coordinate of piece
 * @param {number} y The y coordinate of piece
 * @param {number} grid The grid number of gobang
 * @returns {boolean} Checked result
 */
export const isInBoundOfBoard = (x, y, grid) => x > 0 && x <= grid && y > 0 && y <= grid;
/**
 * Update chess moves
 * @param {array} chessMoves Array of the recordings of chess moves
 * @param {number} x The x coordinate of piece
 * @param {number} y The y coordinate of piece
 * @param {object} props The properties of the component
 * @returns {object} Updated chess moves array, current player name and its color
 */
export const updateChessMoves = (chessMoves, x, y, props) => {
  const { wColor, bColor } = props;
  const nextChessMoves = chessMoves.slice();
  const color = (nextChessMoves.length + 1) % 2 === 0 ? wColor : bColor;
  const player = color === bColor ? black : white;
  nextChessMoves.push({ x, y, color });
  return { nextChessMoves, player, color };
};
/**
 * Translate screen XY coordinates into chessboard XY coordinates
 * @param {element} target The target element of chess pieces
 * @param {number} posX The X screen coordinates
 * @param {number} posY The Y screen coordinates
 * @param {number} grid The grid number of gobang
 * @returns {object} The chessboard XY coordinates
 */
export const calculateXY = (target, posX, posY, grid) => {
  const width = target.getAttribute('width');
  const height = target.getAttribute('height');
  const x = Math.floor(Math.round((posX / width) * (grid + 1)));
  const y = Math.floor(Math.round((posY / height) * (grid + 1)));
  return { x, y };
};
/**
 * Check if the point exists in recordings of chess moves
 * @param {number} x The x coordinate of piece
 * @param {number} y The y coordinate of piece
 * @param {array} chessMoves Array of the recordings of chess moves
 * @param {function} otherConditions Other conditions which want to be added in judgement
 * @returns {boolean} The checking result
 */
export const pointExisted = (x, y, chessMoves, otherConditions = () => true) => {
  const point = chessMoves.find(item => item.x === x && item.y === y && otherConditions(item));
  return point !== undefined;
};
/**
 * Calculate all win conditions
 * @param {number} x The x coordinate of piece
 * @param {number} y The y coordinate of piece
 * @param {number} grid The grid number of gobang
 * @returns {array} All conditions the player could win
 */
export const getWinConditions = (x, y, grid) => {
  const yLowerBound = (y - chainNum) + 1;
  const yUpperBound = (y + chainNum) - 1;
  const xLowerBound = (x - chainNum) + 1;
  const xUPperBound = (x + chainNum) - 1;
  const T = yLowerBound > 0 ?
    Array.from({ length: chainNum }, (item, idx) => [x, y - idx]) :
    undefined;
  const TR = xUPperBound <= grid && yLowerBound > 0 ?
    Array.from({ length: chainNum }, (item, idx) => [x + idx, y - idx]) :
    undefined;
  const R = xUPperBound <= grid ?
    Array.from({ length: chainNum }, (item, idx) => [x + idx, y]) :
    undefined;
  const RB = xUPperBound <= grid && yUpperBound <= grid ?
    Array.from({ length: chainNum }, (item, idx) => [x + idx, y + idx]) :
    undefined;
  const B = yUpperBound <= grid ?
    Array.from({ length: chainNum }, (item, idx) => [x, y + idx]) :
    undefined;
  const LB = xLowerBound > 0 && yUpperBound <= grid ?
    Array.from({ length: chainNum }, (item, idx) => [x - idx, y + idx]) :
    undefined;
  const L = xLowerBound > 0 ?
    Array.from({ length: chainNum }, (item, idx) => [x - idx, y]) :
    undefined;
  const LT = xLowerBound > 0 && yLowerBound > 0 ?
    Array.from({ length: chainNum }, (item, idx) => [x - idx, y - idx]) :
    undefined;
  let result = [T, TR, R, RB, B, LB, L, LT];
  result = result.filter(item => item !== undefined);
  return result;
};
/**
 * Check if the player wins
 * @param {number} x The x coordinate of piece
 * @param {number} y The y coordinate of piece
 * @param {string} color The color of current piece
 * @param {number} grid The grid number of gobang
 * @param {array} chessMoves Array of the recordings of chess moves
 * @returns {boolean} The result if the player wins
 */
export function checkWin(x, y, color, grid, chessMoves) {
  if (chessMoves.length < 9) {
    return false;
  }
  const conditions = getWinConditions(x, y, grid);
  const result = conditions.some(condition => (
    condition.every(point => (
      pointExisted(point[0], point[1], chessMoves, item => item.color === color)
    ))
  ));
  return result;
}
/**
 * Whether game is over
 * @param {array} chessMoves Array of the recordings of chess moves
 * @param {number} x The x coordinate of piece
 * @param {number} y The y coordinate of piece
 * @param {string} color The color of current piece
 * @param {string} player The current player
 * @param {object} props The properties of the component
 * @returns {boolean} If true, the game is over.
 */
export const isGameOver = (chessMoves, x, y, color, player, props) => {
  const { gameOver, grid } = props;
  const result = checkWin(x, y, color, grid, chessMoves);
  if (result) {
    gameOver(player);
  } else if (chessMoves.length === grid * grid) {
    gameOver(deadlock);
  }
  return result;
};
/**
 * Update chess moves if point is valid
 * @param {*} x The x coordinate of piece
 * @param {*} y The y coordinate of piece
 * @param {*} chessMoves Array of the recordings of chess moves
 * @param {*} props The properties of the component
 */
export function updateIfPointIsValid(x, y, chessMoves, props) {
  const { grid } = props;
  if (isInBoundOfBoard(x, y, grid) && !pointExisted(x, y, chessMoves)) {
    const { nextChessMoves, player, color } = updateChessMoves(chessMoves, x, y, props);
    const nextLocked = isGameOver(nextChessMoves, x, y, color, player, props);
    this.setState({ chessMoves: nextChessMoves, locked: nextLocked });
  }
}
/**
 * Add piece on to the chessboard and check if the player wins or deadlocks
 * @param {*} props The properties of the component
 */
export function addPiece(props) {
  return (
    function listener(e) {
      const { grid } = props;
      const { target, offsetX, offsetY } = e;
      const isTarget = target.getAttribute('target');
      const { chessMoves, locked } = this.state;
      if (isTarget && !locked) {
        const { x, y } = calculateXY(target, offsetX, offsetY, grid);
        updateIfPointIsValid.call(this, x, y, chessMoves, props);
      }
    }
  );
}
/**
 * Fallback the piece
 * @param {*} props The properties of the component
 */
export function fallbackPieces(props) {
  const { singleRace } = props;
  const { chessMoves, locked } = this.state;
  const { length } = chessMoves;
  if (!locked) {
    let clonedChessMoves = chessMoves.slice();
    if (singleRace) {
      clonedChessMoves = clonedChessMoves.slice(0, length - 2);
    } else {
      clonedChessMoves = clonedChessMoves.slice(0, length - 1);
    }
    this.setState({ chessMoves: clonedChessMoves });
  }
}
/**
 * Restart the game
 */
export function restartGame() {
  this.setState({ chessMoves: [], locked: false });
}
/**
 * Add click event onto the chessboard
 * @param {*} props The propertis of the component
 */
export function addClickEventListener(props) {
  let listener = addPiece(props);
  listener = listener.bind(this);
  window.addEventListener('click', listener);
  this.setState({ listener });
}
/**
 * Remove click event from chessboard
 */
export function removeClickEventListener() {
  window.removeEventListener('click', this.state.listener);
}
/**
 * Check if fallback or restart is triggered
 * @param {*} nextProps The next properties of the component
 */
export function receivePropsHandler(nextProps) {
  const { fallback, restart } = nextProps;
  if (fallback === true) {
    fallbackPieces.call(this, nextProps);
  }
  if (restart === true) {
    restartGame.call(this);
  }
}
/**
 * Check if the component should be rerendered
 * @param {*} nextState The next state of the component
 * @param {*} state The current state of the component
 */
export const shouldUpdate = (nextState, state) => {
  const { chessMoves: nextChessMoves } = nextState;
  const { chessMoves } = state;
  return nextChessMoves !== chessMoves;
};

export default addClickEventListener;
