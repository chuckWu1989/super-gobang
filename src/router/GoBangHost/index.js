import {
  addClickEventListener,
  removeClickEventListener,
  receivePropsHandler,
  shouldUpdate,
  isCanvasSupported,
} from '../../controllers/GoBangHost';
import GoBangHost from '../../views/GoBangHost';
import { connect } from '../../hocs/connect';

export const method = () => ({
  addClickEventListener,
  removeClickEventListener,
  receivePropsHandler,
  shouldUpdate: (nextState, state) => shouldUpdate(nextState, state),
  isCanvasSupported,
});

export default connect(method)(GoBangHost);
