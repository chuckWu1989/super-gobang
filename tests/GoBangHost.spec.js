import React from 'react';
import toJson from 'enzyme-to-json';
import GoBangHost from '../src/views/GoBangHost';
import { method } from '../src/router/GoBangHost';
import * as controller from '../src/controllers/GoBangHost';
import * as config from '../src/controllers/GoBangHost/config';

const { shallow } = global;
const props = {};

describe('GoBangHost', () => {
  describe('view', () => {
    it('should be rendered properly', () => {
      const ownProps = { ...props };
      const tree = toJson(shallow(<GoBangHost {...ownProps} />));
      expect(tree).toMatchSnapshot();
    });
    it('should call methods during lifecycle', () => {
      const ownProps = {
        ...props,
        addClickEventListener: jest.fn(),
        removeClickEventListener: jest.fn(),
        receivePropsHandler: jest.fn(),
        shouldUpdate: jest.fn(() => true),
        updated: jest.fn(),
      };
      const wrapper = shallow(
        <GoBangHost {...ownProps} />,
        { lifecycleExperimental: true },
      );
      wrapper.setProps({ grid: 14 });
      wrapper.unmount();
      expect(ownProps.addClickEventListener).toHaveBeenCalledTimes(1);
      expect(ownProps.removeClickEventListener).toHaveBeenCalledTimes(1);
      expect(ownProps.receivePropsHandler).toHaveBeenCalledTimes(1);
      expect(ownProps.shouldUpdate).toHaveBeenCalledTimes(1);
      expect(ownProps.updated).toHaveBeenCalledTimes(1);
    });
  });
  describe('router', () => {
    it('should pass addClickEventListener', () => {
      expect('addClickEventListener' in method()).toBeTruthy();
    });
    it('should pass removeClickEventListener', () => {
      expect('removeClickEventListener' in method()).toBeTruthy();
    });
    it('should pass receivePropsHandler', () => {
      expect('receivePropsHandler' in method()).toBeTruthy();
    });
    it('should pass shouldUpdate', () => {
      expect('shouldUpdate' in method()).toBeTruthy();
    });
  });
  describe('controller', () => {
    describe('isInBoundOfBoard', () => {
      const grid = 15;
      it('should return false if x <= 0', () => {
        const x = 0;
        const y = 1;
        const result = controller.isInBoundOfBoard(x, y, grid);
        expect(result).toBeFalsy();
      });
      it('should return false if x > grid', () => {
        const x = 16;
        const y = 1;
        const result = controller.isInBoundOfBoard(x, y, grid);
        expect(result).toBeFalsy();
      });
      it('should return false if y <= 0', () => {
        const x = 1;
        const y = 0;
        const result = controller.isInBoundOfBoard(x, y, grid);
        expect(result).toBeFalsy();
      });
      it('should return false if y > grid', () => {
        const x = 1;
        const y = 16;
        const result = controller.isInBoundOfBoard(x, y, grid);
        expect(result).toBeFalsy();
      });
      it('should return true if 0 <= x <= grid and 0 <= y <= grid', () => {
        const x = 1;
        const y = 1;
        const result = controller.isInBoundOfBoard(x, y, grid);
        expect(result).toBeTruthy();
      });
    });
    describe('updateChessMoves', () => {
      const x = 1;
      const y = 1;
      const ownProps = { ...props, wColor: '#ffffff', bColor: '#000000' };
      it('should update chess moves with black piece and black player if the length of updated chess is odd', () => {
        const chessmoves = [];
        const result = controller.updateChessMoves(chessmoves, x, y, ownProps);
        expect(result.nextChessMoves.length).toBe(1);
        expect(result.player).toBe(config.black);
        expect(result.color).toBe(ownProps.bColor);
      });
      it('should update chess moves with white piece and white player if the length of updated chess is even', () => {
        const chessmoves = [{ x: 10, y: 10, color: '#000000' }];
        const result = controller.updateChessMoves(chessmoves, x, y, ownProps);
        expect(result.nextChessMoves.length).toBe(2);
        expect(result.player).toBe(config.white);
        expect(result.color).toBe(ownProps.wColor);
      });
    });
    describe('calculateXY', () => {
      it('should return coordinate x and y', () => {
        const target = {
          getAttribute: () => 160,
        };
        const posX = 14;
        const posY = 14;
        const grid = 15;
        const result = controller.calculateXY(target, posX, posY, grid);
        expect(result.x).toBe(1);
        expect(result.y).toBe(1);
      });
    });
    describe('pointExisted', () => {
      it('should return true if point is defined in chess moves list', () => {
        const x = 10;
        const y = 10;
        const color = '#000000';
        const chessMoves = [{ x, y, color }];
        const otherConditions = ({ color: itemColor }) => itemColor === color;
        const result = controller.pointExisted(x, y, chessMoves, otherConditions);
        expect(result).toBeTruthy();
      });
      it('should return false if point is undefined in chess moves list', () => {
        const x = 10;
        const y = 10;
        const chessMoves = [{ x, y }];
        const result = controller.pointExisted(11, 11, chessMoves);
        expect(result).toBeFalsy();
      });
    });
    describe('getWinConditions', () => {
      const grid = 15;
      it('should get 8 win conditions if x directional chain and y directional chain are in region', () => {
        const x = 7;
        const y = 7;
        const result = controller.getWinConditions(x, y, grid);
        expect(result.length).toBe(8);
      });
      it('should get 5 win conditions if y directional chain exceeds top boundary', () => {
        const x = 7;
        const y = 1;
        const result = controller.getWinConditions(x, y, grid);
        expect(result.length).toBe(5);
      });
      it('should get 3 win conditions if y directional chain exceeds top boundary and x directional chain exceeds right boundary', () => {
        const x = 13;
        const y = 1;
        const result = controller.getWinConditions(x, y, grid);
        expect(result.length).toBe(3);
      });
      it('should get 5 win conditions if x directional chain exceeds right boundary', () => {
        const x = 13;
        const y = 7;
        const result = controller.getWinConditions(x, y, grid);
        expect(result.length).toBe(5);
      });
      it('should get 3 win conditions if x directional chain exceeds right boundary and y directional chain exceeds bottom boundary', () => {
        const x = 13;
        const y = 13;
        const result = controller.getWinConditions(x, y, grid);
        expect(result.length).toBe(3);
      });
      it('should get 5 win conditions if y directional chain exceeds bottom boundary', () => {
        const x = 7;
        const y = 13;
        const result = controller.getWinConditions(x, y, grid);
        expect(result.length).toBe(5);
      });
      it('should get 3 win conditions if x directional chain exceeds left boundary and y directional chain exceeds bottom boundary', () => {
        const x = 3;
        const y = 13;
        const result = controller.getWinConditions(x, y, grid);
        expect(result.length).toBe(3);
      });
      it('should get 5 win conditions if x directional chain exceeds left boundary', () => {
        const x = 3;
        const y = 7;
        const result = controller.getWinConditions(x, y, grid);
        expect(result.length).toBe(5);
      });
      it('should get 3 win conditions if x directional chain exceeds left boundary and y directional chain exceeds top boundary', () => {
        const x = 3;
        const y = 3;
        const result = controller.getWinConditions(x, y, grid);
        expect(result.length).toBe(3);
      });
    });
    describe('checkWin', () => {
      const x = 5;
      const y = 1;
      const grid = 15;
      it('should return true if chess moves match one of win condition', () => {
        const chessmoves = [
          { x: 1, y: 1, color: config.bColor },
          { x: 1, y: 2, color: config.wColor },
          { x: 2, y: 1, color: config.bColor },
          { x: 2, y: 2, color: config.wColor },
          { x: 3, y: 1, color: config.bColor },
          { x: 3, y: 2, color: config.wColor },
          { x: 4, y: 1, color: config.bColor },
          { x: 4, y: 2, color: config.wColor },
          { x: 5, y: 1, color: config.bColor },
        ];
        const result = controller.checkWin(x, y, config.bColor, grid, chessmoves);
        expect(result).toBeTruthy();
      });
      it('should return false if the length of chess moves is lower than 9', () => {
        const chessmoves = [];
        const result = controller.checkWin(x, y, config.bColor, grid, chessmoves);
        expect(result).toBeFalsy();
      });
    });
    describe('isGameOver', () => {
      it('should return true and call gameover with winner if checkWin said there is a winner', () => {
        let winner;
        const chessmoves = [
          { x: 1, y: 1, color: config.bColor },
          { x: 1, y: 2, color: config.wColor },
          { x: 2, y: 1, color: config.bColor },
          { x: 2, y: 2, color: config.wColor },
          { x: 3, y: 1, color: config.bColor },
          { x: 3, y: 2, color: config.wColor },
          { x: 4, y: 1, color: config.bColor },
          { x: 4, y: 2, color: config.wColor },
          { x: 5, y: 1, color: config.bColor },
        ];
        const { x, y, color } = chessmoves[chessmoves.length - 1];
        const player = config.black;
        const ownProps = { gameOver: jest.fn((result) => { winner = result; }), grid: 15 };
        const result = controller.isGameOver(chessmoves, x, y, color, player, ownProps);
        expect(result).toBeTruthy();
        expect(winner).toBe(player);
      });
      it('should return false and call gameover with deadlock if full pieces on chess board and there is no winner', () => {
        let winner;
        const chessmoves = Array.from(
          { length: 225 },
          () => ({ x: 1, y: 1, color: config.wColor }),
        );
        const x = 16;
        const y = 16;
        const color = config.bColor;
        const player = config.black;
        const ownProps = { gameOver: jest.fn((result) => { winner = result; }), grid: 15 };
        const result = controller.isGameOver(chessmoves, x, y, color, player, ownProps);
        expect(result).toBeFalsy();
        expect(winner).toBe(config.deadlock);
      });
      it('should return false if checkWin said there is no winner', () => {
        const chessmoves = [
          { x: 1, y: 1, color: config.bColor },
          { x: 1, y: 2, color: config.wColor },
          { x: 2, y: 1, color: config.bColor },
          { x: 2, y: 2, color: config.wColor },
          { x: 3, y: 1, color: config.bColor },
          { x: 3, y: 2, color: config.wColor },
          { x: 4, y: 1, color: config.bColor },
          { x: 4, y: 2, color: config.wColor },
        ];
        const { x, y, color } = chessmoves[chessmoves.length - 1];
        const player = config.white;
        const ownProps = { gameOver: jest.fn(), grid: 15 };
        const result = controller.isGameOver(chessmoves, x, y, color, player, ownProps);
        expect(result).toBeFalsy();
      });
    });
    describe('updateIfPointIsValid', () => {
      it('should update piece into chess moves if the piece doesnt exist and its in boundary', () => {
        let result;
        const x = 7;
        const y = 7;
        const chessmoves = [];
        const ownProps = { ...props, grid: 15 };
        const self = { setState: (response) => { result = response; } };
        controller.updateIfPointIsValid.call(self, x, y, chessmoves, ownProps);
        expect(result.chessMoves.length).toBe(1);
      });
      it('should skip updating if the piece is already on the board', () => {
        let result;
        const chessmoves = [{ x: 7, y: 7, color: config.bColor }];
        const { x, y } = chessmoves[0];
        const ownProps = { ...props, grid: 15 };
        const self = { setState: (response) => { result = response; } };
        controller.updateIfPointIsValid.call(self, x, y, chessmoves, ownProps);
        expect(result).toBeUndefined();
      });
    });
    describe('addPiece', () => {
      const ownProps = {
        ...props,
        bColor: config.bColor,
        wColor: config.wColor,
        grid: 15,
      };
      const e = {
        target: {
          getAttribute: (type) => {
            switch (type) {
              case 'target': return true;
              case 'width': return 160;
              case 'height': return 160;
              default: break;
            }
            return undefined;
          },
        },
        offsetX: 14,
        offsetY: 14,
      };
      const self = {
        setState: jest.fn(),
        state: {
          locked: false,
          chessMoves: [],
        },
      };
      it('should return hanlder function', () => {
        const result = controller.addPiece();
        expect(typeof result).toBe('function');
      });
      it('should update pieces if target is right and locked is false', () => {
        controller.addPiece(ownProps).call(self, e);
        expect(self.setState).toHaveBeenCalledTimes(1);
      });
      it('should skip updating if target is wrong or locked is true', () => {
        self.state.locked = true;
        self.setState = jest.fn();
        controller.addPiece(ownProps).call(self, e);
        expect(self.setState).not.toHaveBeenCalled();
      });
    });
    describe('fallbackPieces', () => {
      it('should fallback 1 piece if locked is false and single mode is false', () => {
        let result;
        const ownProps = { ...props, singleRace: false };
        const self = {
          setState: (response) => { result = response; },
          state: {
            chessMoves: [{ x: 1, y: 1, color: config.bColor }],
            locked: false,
          },
        };
        controller.fallbackPieces.call(self, ownProps);
        expect(result.chessMoves.length).toBe(0);
      });
      it('should fallback 2 pieces if locked is false and single mode is true', () => {
        let result;
        const ownProps = { ...props, singleRace: true };
        const self = {
          setState: (response) => { result = response; },
          state: {
            chessMoves: [
              { x: 1, y: 1, color: config.bColor },
              { x: 1, y: 2, color: config.wColor },
            ],
            locked: false,
          },
        };
        controller.fallbackPieces.call(self, ownProps);
        expect(result.chessMoves.length).toBe(0);
      });
      it('should skip if locked is true', () => {
        let result;
        const ownProps = { ...props, singleRace: true };
        const self = {
          setState: (response) => { result = response; },
          state: {
            chessMoves: [{ x: 1, y: 1, color: config.bColor }],
            locked: true,
          },
        };
        controller.fallbackPieces.call(self, ownProps);
        expect(result).toBeUndefined();
      });
    });
    describe('restartGame', () => {
      it('should reset state', () => {
        let result;
        const self = { setState: (response) => { result = response; } };
        controller.restartGame.call(self);
        expect(result.chessMoves.length).toBe(0);
        expect(result.locked).toBeFalsy();
      });
    });
    describe('addClickEventListener', () => {
      it('should add click event listener and set it as state', () => {
        let result;
        const temp = window.addEventListener;
        window.addEventListener = jest.fn();
        const self = { setState: (response) => { result = response; } };
        controller.addClickEventListener.call(self);
        expect(window.addEventListener).toHaveBeenCalledTimes(1);
        expect('listener' in result).toBeTruthy();
        window.addEventListener = temp;
      });
    });
    describe('removeClickEventListener', () => {
      it('should call removeEventlistener', () => {
        const temp = window.removeEventListener;
        window.removeEventListener = jest.fn();
        const self = { state: { listener: () => {} } };
        controller.removeClickEventListener.call(self);
        expect(window.removeEventListener).toHaveBeenCalledTimes(1);
        window.removeEventListener = temp;
      });
    });
    describe('receivePropsHandler', () => {
      it('should fallback piece if fallback is true', () => {
        const ownProps = {
          ...props,
          fallback: true,
          restart: false,
          singleRace: false,
        };
        const self = {
          setState: jest.fn(),
          state: {
            chessMoves: [{ x: 1, y: 1, color: config.bColor }],
            locked: false,
          },
        };
        controller.receivePropsHandler.call(self, ownProps);
        expect(self.setState).toHaveBeenCalledTimes(1);
      });
      it('should restart game if restart is true', () => {
        const ownProps = { ...props, fallback: false, restart: true };
        const self = { setState: jest.fn() };
        controller.receivePropsHandler.call(self, ownProps);
        expect(self.setState).toHaveBeenCalledTimes(1);
      });
      it('should skip if restart and fallback both false', () => {
        const ownProps = { ...props, fallback: false, restart: false };
        const self = { setState: jest.fn() };
        controller.receivePropsHandler.call(self, ownProps);
        expect(self.setState).toHaveBeenCalledTimes(0);
      });
    });
    describe('shouldUpdate', () => {
      it('should return true if the length of next chess moves is different with current one', () => {
        const nextState = { chessMoves: [{ x: 1, y: 1, color: config.bColor }] };
        const state = { chessMoves: [] };
        const result = controller.shouldUpdate(nextState, state);
        expect(result).toBeTruthy();
      });
      it('should return false if the length of next chess moves is same with current one', () => {
        const nextState = { chessMoves: [] };
        const state = nextState;
        const result = controller.shouldUpdate(nextState, state);
        expect(result).toBeFalsy();
      });
    });
  });
});
