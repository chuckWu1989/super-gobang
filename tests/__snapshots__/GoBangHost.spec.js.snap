// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GoBangHost view should be rendered properly 1`] = `
<div
  className="gobang-host-style"
>
  <Component>
    <Component
      chessMoves={Array []}
      grid={15}
    />
  </Component>
</div>
`;
